pkg/nmi/server/server.go

```diff
@@ -70,12 +96,14 @@ func NewServer(micNamespace string, blockInstanceMetadata bool) *Server {
 		MICNamespace:          micNamespace,
 		BlockInstanceMetadata: blockInstanceMetadata,
 		Reporter:              reporter,
+		PodClient:             podClient,
+		PodObjChannel:         podObjCh,
 	}
 }
 
-// Run runs the specified Server.
-func (s *Server) Run() error {
-	go s.updateIPTableRules()
+// Run the specified Server.
+func (s *Server) Run(r RedirectorInt) error {
+	r.RedirectMetadataEndpoint()
```

Comment:
>r.RedirectMetadataEndpoint() [](start = 1, length = 28)

I do not see much use for this interface to be passed to a server - so it can call the redirector in the first line and forget about it.
It hides lot of important internal details about how redirector works and this does not help with maintaining the code in long run.

I would separate running redirector and running server into two independent parts. There is no reason to bunch them together.

Correctly designed interface will provide better answers to following questions:
* It is not clear how this redirector runs - is it one shot or a goroutine?
* It is not clear how redirector is parameterized - what parameters it depends on (needs)?
* If this redirector is a goroutine - when how it is terminated, is there any cleanup needed during termination?

Thoughts:
* Do we need a Redirector interface - in its current form a function type would work just as fine

Reply:
Platform specific implementations for `redirector_<platform>.go` should contain both methods (WindowsRedirector and LinuxRedirector) but one of them that is not for this platform should just panic

---
In reply to: [383488591](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r383488591) [](ancestors = 383488591)

Reply:
Update the code as comments.